src/modem.rs

```diff
@@ -0,0 +1,752 @@
+use crate::handle::RawHandle;
+use core::{borrow::BorrowMut, ffi, marker::PhantomData, mem::MaybeUninit};
+use esp_idf_hal::{
+    delay::BLOCK,
+    uart::{UartDriver, UartTxDriver},
+};
+use std::sync::Arc;
+
+use crate::{
+    eventloop::{
+        EspEventDeserializer, EspEventLoop, EspEventSource, EspSubscription, EspSystemEventLoop,
+        System,
+    },
+    netif::{EspNetif, EspNetifDriver, NetifStack},
+    private::mutex,
+    sys::*,
+};
+
+pub struct EspModem<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>>,
+{
+    serial: T,
+    status: Arc<mutex::Mutex<ModemDriverStatus>>,
+    _subscription: EspSubscription<'static, System>,
+    netif: EspNetif,
+    _d: PhantomData<&'d ()>,
+}
+
+impl<'d, T> EspModem<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>> + Send,
+{
+    pub fn new(serial: T, sysloop: EspSystemEventLoop) -> Result<Self, EspError> {
+        let (status, subscription) = Self::subscribe(&sysloop)?;
+
+        Ok(Self {
+            serial,
+            status,
+            _subscription: subscription,
+            netif: EspNetif::new(NetifStack::Ppp)?,
+            _d: PhantomData,
+        })
+    }
+
+    /// Run the modem network interface. Blocks until the PPP encounters an error.
+    pub fn run(&mut self, mut buffer: [u8; 64]) -> Result<(), EspError> {
```

Comment:
Nit: `buffer: &mut [u8]`. (a) you probably don't want to move the buffer into `run` (b) why constrain the size of the buffer to 64 bytes? User could pass whatever she wants.

Reply:
Yeah done